src/peerDiscovery.cpp

```diff
@@ -0,0 +1,378 @@
+
+    m_port_received = PeerDiscovery::litendtoint(data_port);
+    m_node_id_received = dht::InfoHash(data_infohash, 20);
+
+}
+
+void
+PeerDiscovery::Sender_Setup(dht::InfoHash nodeId, in_port_t port_to_send){
+
+    //Set up for Sender
+    initialize_socket();
+    initialize_sockaddr_Sender();
+
+    //Setup for send data
+    uint8_t *x = nodeId.data();
+    int port_node = port_to_send;
+    uint8_t port_node_binary[2];
+    PeerDiscovery::inttolitend(port_node,port_node_binary);
+
+    //Copy Node id and node port
+    memcpy (m_data_send, x, 20);
+    m_data_send[20] = port_node_binary[0];
+    m_data_send[21] = port_node_binary[1];
+    
+}
+
+void
+PeerDiscovery::SenderThread(bool &continues){
+
+    while(continues){
+        
+        m_sendto(m_data_send,m_data_size);
+        sleep(5);
+
+    }
+
+}
+
+void
+PeerDiscovery::Sender_Loop(){
+
+    m_running = std::thread(&PeerDiscovery::SenderThread, this, std::ref(m_continue_to_run));
+
+}
+
+void
+PeerDiscovery::Listener_Setup(){
+
+    initialize_socket();
+    initialize_socketopt();
+    initialize_sockaddr_Listener();
+    mcast_join();
+
+}
+
+void
+PeerDiscovery::Listener_Thread(PeerDiscoveredCallback callback, bool &continues){
+
+#ifndef _WIN32
+    auto status = pipe(m_stopfds_pipe);
+    if (status == -1) {
+        throw DhtException(std::string("Can't open pipe: ") + strerror(errno));
+    }
+#else
+    udpPipe(stopfds);
+#endif
+    m_stop_readfd = m_stopfds_pipe[0];
+    m_stop_writefd = m_stopfds_pipe[1];
+
+    while(continues){
+        
+        fd_set readfds;
+
+        FD_ZERO(&readfds);
+        FD_SET(m_stop_readfd, &readfds);
+        FD_SET(m_sockfd, &readfds);
+
+        int data_coming = select(m_sockfd + 1, &readfds, nullptr, nullptr, nullptr);
+        if(data_coming < 0) {
+            if(errno != EINTR) {
+                perror("select");
+                std::this_thread::sleep_for( std::chrono::seconds(1) );
+            }
+        }
+
+        if(data_coming > 0){
+            
+            if(FD_ISSET(m_stop_readfd,&readfds)){ break; }
+
+            size_t data_receive_size = 22;
+            uint8_t data_receive[22];
+
+            m_recvfrom(data_receive,data_receive_size);
+
+            uint8_t data_infohash[20];
+            uint8_t data_port[2];
+
+            memcpy (data_infohash, data_receive, 20);
+            data_port[0] = data_receive[20];
+            data_port[1] = data_receive[21];
+
+            auto port = PeerDiscovery::litendtoint(data_port);
+            auto nodeId = dht::InfoHash(data_infohash, 20);
+
+            if(m_port_self != port){
+
+                m_sockaddr.setPort(port);
+                callback(nodeId, m_sockaddr);
+
+            }
+            sleep(1);
+
+        }
+
+    }
+    if (m_stop_readfd != -1)
+        close(m_stop_readfd);
+    if (m_stop_writefd != -1)
+        close(m_stop_writefd);
+
+}
+
+void 
+PeerDiscovery::Listener_Loop(PeerDiscoveredCallback callback)
```

Comment:
dedicated method not needed for this. Thread can be instantiated in startDiscovery or in Listener_Thread